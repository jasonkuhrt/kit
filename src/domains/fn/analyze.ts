import { Lang } from '#lang'

/**
 * @category Introspection
 */
export type Parameter = { type: 'name'; value: string } | { type: 'destructured'; names: string[] }

/**
 * Analyze a function to extract its parameter information and body.
 *
 * Parses a function's string representation to extract:
 * - Parameter names (both regular and destructured parameters)
 * - Function body (both statement and expression forms, trimmed and dedented)
 *
 * The returned body is already cleaned: leading/trailing whitespace removed and
 * common indentation stripped away for clean display in its isolated form.
 *
 * @category Introspection
 * @param fn - The function to analyze
 * @returns An object containing the function's cleaned body and parameters
 * @throws {Error} If the function cannot be parsed or has invalid structure
 *
 * @example
 * ```ts
 * const fn = (a, { b, c }) => a + b + c
 * const info = analyzeFunction(fn)
 * // info.parameters: [{ type: 'name', value: 'a' }, { type: 'destructured', names: ['b', 'c'] }]
 * // info.body: "a + b + c" (already trimmed and dedented)
 * ```
 */
export const analyzeFunction = (fn: (...args: [...any[]]) => unknown) => {
  const groups = fn.toString().match(functionPattern)?.groups
  if (!groups) throw new Error(`Could not extract groups from function.`)

  const bodyRaw = groups[`bodyStatement`] ?? groups[`bodyExpression`]
  if (bodyRaw === undefined) throw new Error(`Could not extract body from function.`)

  // Clean the body: strip common indentation first, then trim
  // Inlined stripIndent logic to avoid circular dependency with str module
  const bodyLines = bodyRaw.split('\n')
  const indents = bodyLines
    .filter((line) => line.trim().length > 0)
    .map((line) => {
      const match = line.match(/^(\s*)/)
      return match?.[1]?.length ?? 0
    })

  const minIndent = indents.length > 0 ? Math.min(...indents) : 0
  const dedented = bodyLines.map((line) => line.slice(minIndent)).join('\n')
  const body = dedented.trim()

  const parameters: Parameter[] = []

  if (groups[`parameters`]) {
    const results = [...groups[`parameters`].matchAll(functionParametersPattern)]
    const resultParameters = results.map(result => {
      const type = result.groups?.[`destructured`] ? `destructured` : result.groups?.[`name`] ? `name` : null

      switch (type) {
        case `destructured`:
          const valueRaw = result.groups![`destructured`]!
          const names = [...valueRaw.matchAll(destructuredPattern)].map(result => {
            const name = result.groups![`name`]
            if (name === undefined) throw new Error(`Could not extract name from destructured parameter.`)
            return name
          })
          return {
            type,
            names,
          } as const
        case `name`:
          return {
            type,
            value: result.groups![`name`]!,
          } as const
        case null:
          throw new Error(`Could not determine type of parameter.`)
        default:
          throw Lang.neverCase(type)
      }
    })

    parameters.push(...resultParameters)
  }

  return {
    body,
    parameters,
  }
}

/**
 * @see https://regex101.com/r/U0JtfS/1
 */
const functionPattern =
  /^(?:(?<async>async)\s+)?(?:function\s+)?(?:(?<name>[A-z_0-9]+)\s*)?\((?<parameters>[^)]*)\)\s*(?:=>\s*(?<bodyExpression>[^\s{].*)|(?:=>\s*)?{(?<bodyStatement>.*)})$/s

/**
 * @see https://regex101.com/r/tE2dV5/2
 */
const functionParametersPattern = /(?<destructured>\{[^}]+\})|(?<name>[A-z_][A-z_0-9]*)/gs

/**
 * https://regex101.com/r/WHwazx/1
 */
const destructuredPattern = /(?<name>[A-z_][A-z_0-9]*)(?::[^},]+)?/gs
