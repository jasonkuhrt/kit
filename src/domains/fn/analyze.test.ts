import { Test } from '#test'
import { analyzeFunction } from './analyze.js'

Test.on(analyzeFunction)
  .describe('parameters - named')
  .casesAsArgs(
    (a: number) => a,
    (a: number, b: string, c: boolean) => [a, b, c],
    () => 42,
    (_a: number, __b: string) => _a,
    (...args: any[]) => args,
  )
  .describe('parameters - destructured')
  .casesAsArgs(
    ({ a, b }: { a: number; b: string }) => [a, b],
    (a: number, { b, c }: { b: string; c: boolean }) => [a, b, c],
  )
  .describe('body extraction')
  .casesAsArgs(
    (a: number) => a + 1,
    (a: number) => {
      return a + 1
    },
    function f(a: number) {
      return a + 1
    },
    (a: number) => {
      const b = a + 1
      const c = b * 2
      return c
    },
  )
  .describe('async functions')
  .casesAsArgs(
    async (a: number) => a + 1,
    async (a: number) => {
      return a + 1
    },
    async function f(a: number) {
      return a + 1
    },
  )
  .describe('error cases')
  .casesAsArgs(
    'not a function' as any, // Error will be caught and snapshotted as "THEN THROWS"
  )
  .test()
