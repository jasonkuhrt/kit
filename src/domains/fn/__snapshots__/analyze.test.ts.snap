// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`async functions > analyzeFunction(async (a) => { return a + 1; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
async (a) => {
    return a + 1;
  }
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'return a + 1;',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`async functions > analyzeFunction(async (a) => a + 1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
async (a) => a + 1
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'a + 1',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`async functions > analyzeFunction(async function f2(a) { return a + 1; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
async function f2(a) {
    return a + 1;
  }
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'return a + 1;',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`body extraction > analyzeFunction((a) => { const b = a + 1; const c = b * 2; return c; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a) => {
    const b = a + 1;
    const c = b * 2;
    return c;
  }
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'const b = a + 1;\\nconst c = b * 2;\\nreturn c;',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`body extraction > analyzeFunction((a) => { return a + 1; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a) => {
    return a + 1;
  }
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'return a + 1;',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`body extraction > analyzeFunction((a) => a + 1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a) => a + 1
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'a + 1',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`body extraction > analyzeFunction(function f(a) { return a + 1; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
function f(a) {
    return a + 1;
  }
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'return a + 1;',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`error cases > analyzeFunction('not a function') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
not a function
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Could not extract groups from function.
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - destructured > analyzeFunction(({ a, b }) => [a, b]) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
({ a, b }) => [a, b]
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: '[a, b]',
  parameters: [
    {
      type: 'destructured',
      names: [ 'a', 'b' ]
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - destructured > analyzeFunction((a, { b, c }) => [a, b, c]) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a, { b, c }) => [a, b, c]
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: '[a, b, c]',
  parameters: [
    {
      type: 'name',
      value: 'a'
    },
    {
      type: 'destructured',
      names: [ 'b', 'c' ]
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - named > analyzeFunction(() => 42) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
() => 42
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: '42',
  parameters: []
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - named > analyzeFunction((...args) => args) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(...args) => args
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'args',
  parameters: [
    {
      type: 'name',
      value: 'args'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - named > analyzeFunction((_a, __b) => _a) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(_a, __b) => _a
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: '_a',
  parameters: [
    {
      type: 'name',
      value: '_a'
    },
    {
      type: 'name',
      value: '__b'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - named > analyzeFunction((a) => a) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a) => a
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: 'a',
  parameters: [
    {
      type: 'name',
      value: 'a'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;

exports[`parameters - named > analyzeFunction((a, b, c) => [a, b, c]) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
(a, b, c) => [a, b, c]
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  body: '[a, b, c]',
  parameters: [
    {
      type: 'name',
      value: 'a'
    },
    {
      type: 'name',
      value: 'b'
    },
    {
      type: 'name',
      value: 'c'
    }
  ]
}
╚══════════════════════════════════════════════════╝"
`;
